packages/bus-query-operation/lib/Bindings.ts

```diff
@@ -50,3 +52,195 @@ export function isBindings(maybeBindings: any): boolean {
+ * @return Algebra.Operation A new operation materialized with the given bindings.
+ */
+export function materializeOperation(operation: Algebra.Operation, bindings: Bindings,
+                                     strictTargetVariables: boolean = false): Algebra.Operation {
+  return Util.mapOperation(operation, {
+    path: (op: Algebra.Path, factory: Factory) => {
+      // Materialize variables in a path expression.
+      // The predicate expression will be recursed.
+      return {
+        recurse: false,
+        result: factory.createPath(
+          materializeTerm(op.subject, bindings),
+          op.predicate,
+          materializeTerm(op.object, bindings),
+          materializeTerm(op.graph, bindings),
+        ),
+      };
+    },
+    pattern: (op: Algebra.Pattern, factory: Factory) => {
+      // Materialize variables in the quad pattern.
+      return {
+        recurse: false,
+        result: factory.createPattern(
+          materializeTerm(op.subject, bindings),
+          materializeTerm(op.predicate, bindings),
+          materializeTerm(op.object, bindings),
+          materializeTerm(op.graph, bindings),
+        ),
+      };
+    },
+    extend: (op: Algebra.Extend, factory: Factory) => {
+      // Materialize an extend operation.
+      // If strictTargetVariables is true, we throw if the extension target variable is attempted to be bound.
+      // Otherwise, we bind the variable to the bound value directly.
+      if (bindings.has(termToString(op.variable))) {
+        if (strictTargetVariables) {
+          throw new Error(`Tried to bind variable ${termToString(op.variable)} in a BIND operator.`);
+        } else {
+          return {
+            recurse: true,
+            result: factory.createExtend(
+              op.input,
+              op.variable,
+              factory.createTermExpression(bindings.get(termToString(op.variable)))
+            ),
+          }
+        }
+      }
+      return {
+        recurse: true,
+        result: op,
+      };
+    },
+    group: (op: Algebra.Group, factory: Factory) => {
+      // Materialize a group operation.
+      // If strictTargetVariables is true, we throw if the group target variable is attempted to be bound.
+      // Otherwise, we just filter out the bound variables.
+      if (strictTargetVariables) {
+        for (const variable of op.variables) {
+          if (bindings.has(termToString(variable))) {
+            throw new Error(`Tried to bind variable ${termToString(variable)} in a GROUP BY operator.`);
+          }
+        }
+        return {
+          recurse: true,
+          result: op,
+        };
+      } else {
+        const variables = op.variables.filter((variable) => !bindings.has(termToString(variable)));
+        return {
+          recurse: true,
+          result: factory.createGroup(
+            op.input,
+            variables,
+            op.aggregates,
+          ),
+        };
```

Comment:
This could create problems further down the line if variables is empty but I guess we don't care much if users break their own queries?

Reply:
Yes indeed, I think that's the safest way to do it.